synapse/util/async_helpers.py

```diff
@@ -428,56 +414,72 @@ def _ctx_manager() -> Iterator[None]:
+                        self.name,
+                        key,
+                    )
+                else:
+                    logger.warning(
+                        "Unexpected exception waiting for linearizer lock %r for key "
+                        "%r",
+                        self.name,
+                        key,
+                    )
 
-        try:
-            await new_defer
-        except Exception as e:
-            logger.info("defer %r got err %r", new_defer, e)
-            if isinstance(e, CancelledError):
-                logger.debug(
-                    "Cancelling wait for linearizer lock %r for key %r", self.name, key
-                )
+                # we just have to take ourselves back out of the queue.
+                del entry.deferreds[new_defer]
+                raise
 
-            else:
-                logger.warning(
-                    "Unexpected exception waiting for linearizer lock %r for key %r",
-                    self.name,
-                    key,
-                )
-
-            # we just have to take ourselves back out of the queue.
-            del entry.deferreds[new_defer]
-            raise
-
-        logger.debug("Acquired linearizer lock %r for key %r", self.name, key)
-        entry.count += 1
-
-        # if the code holding the lock completes synchronously, then it
-        # will recursively run the next claimant on the list. That can
-        # relatively rapidly lead to stack exhaustion. This is essentially
-        # the same problem as http://twistedmatrix.com/trac/ticket/9304.
-        #
-        # In order to break the cycle, we add a cheeky sleep(0) here to
-        # ensure that we fall back to the reactor between each iteration.
-        #
-        # (This needs to happen while we hold the lock, and the context manager's exit
-        # code must be synchronous, so this is the only sensible place.)
-        await self._clock.sleep(0)
+            logger.debug("Acquired linearizer lock %r for key %r", self.name, key)
+            entry.count += 1
+
+            # if the code holding the lock completes synchronously, then it
+            # will recursively run the next claimant on the list. That can
+            # relatively rapidly lead to stack exhaustion. This is essentially
+            # the same problem as http://twistedmatrix.com/trac/ticket/9304.
+            #
+            # In order to break the cycle, we add a cheeky sleep(0) here to
+            # ensure that we fall back to the reactor between each iteration.
+            #
+            # (This needs to happen while we hold the lock, and the context manager's
+            # exit code must be synchronous, so this is the only sensible place.)
+            await self._clock.sleep(0)
+        else:
+            logger.debug(
+                "Acquired uncontended linearizer lock %r for key %r", self.name, key
+            )
+            entry.count += 1
```

Comment:
it'd be nice to do an early return here, rather than have a big `if` block which is hard to follow.